### PR TITLE
Fix device creation

### PR DESCRIPTION
**Pull request by ghost**
- allow users to enter a device name
- fix field name for monitoring url

### CODE REVIEW

**Comment by sfeilmeier** [comment]
@slash-ds: We are not anymore supporting Odoo v15, since we updated to v16. See https://github.com/OpenEMS/odoo-openems/tree/16.0

Is this still relevant to you?

**Comment by ghost** [comment]
@sfeilmeier I think it is still relevant if I have not missed anything. 
- The key for the configuration parameter "edge_monitoring_url" is wrong
- the required name for device is readonly --> no creation is possible via UI
- the computed field is not calculated correctly (TypeError: can only concatenate str (not "bool") to str)

I change the target branch to "16.0".

**Comment by miettal** [comment]
Exactly I had plan to send almost same content pull-request!!

odoo-openems 16.0 has problem still, same to 15.0.

↓ is my patch.
I tested ↓ patch on odoo 15 and 16 and it works. so @slash-ds -san's PR will work in odoo 16 I think.

```
diff --git a/openems/models/device.py b/openems/models/device.py
index ce22fc4..c764375 100644
--- a/openems/models/device.py
+++ b/openems/models/device.py
@@ -41,8 +41,9 @@ class Device(models.Model):
     @api.depends("name")
     def _compute_monitoring_url(self):
         for rec in self:
-            url = self.env["ir.config_parameter"].sudo().get_param("openems.edge_monitoring_url", default='#')
-            rec.monitoring_url = url + rec.name
+            url = self.env["ir.config_parameter"].sudo().get_param("edge_monitoring_url", default='#')
+            name = rec.name if rec.name is not False else ''
+            rec.monitoring_url = url + name

     producttype = fields.Selection(
         [
diff --git a/openems/views/device.xml b/openems/views/device.xml
index cf11526..84c72ae 100644
--- a/openems/views/device.xml
+++ b/openems/views/device.xml
@@ -80,7 +80,7 @@
                         <h1>
                             <field name="comment" nolabel="True" />
                         </h1>
-                        <field name="name" readonly="1" />
+                        <field name="name" />
                         <field name="monitoring_url" widget="url" />
                     </group>
                     <notebook>
```

**Comment by miettal** [comment]
I confirmed this behavior too. I couldn't add new edge on Odoo UI. // I just setupped new odoo instance without other app installation and customization.

> The key for the configuration parameter "edge_monitoring_url" is wrong
> the required name for device is readonly --> no creation is possible via UI
> the computed field is not calculated correctly (TypeError: can only concatenate str (not "bool") to str)

File: openems/models/device.py [hunk 1]
```diff
@@ -42,7 +42,7 @@ def _compute_first_setup_protocol(self):
     def _compute_monitoring_url(self):
         for rec in self:
             url = self.env["ir.config_parameter"].sudo().get_param("openems.edge_monitoring_url", default='#')
-            rec.monitoring_url = url + rec.name
+            rec.monitoring_url =  (url + rec.name) if rec.name else ""
```
Comment (sfeilmeier) [inline openems/models/device.py:45]: This is not required because of `default='#'` above

File: openems/views/device.xml [hunk 2]
```diff
@@ -80,7 +80,7 @@
                         <h1>
                             <field name="comment" nolabel="True" />
                         </h1>
-                        <field name="name" readonly="1" />
+                        <field name="name" />
```
Comment (sfeilmeier) [inline openems/views/device.xml:83]: See my comments below, for why I do not fully like this solution.

**Review by sfeilmeier — COMMENTED** [review]
> I confirmed this behavior too. I couldn't add new edge on Odoo UI. // I just setupped new odoo instance without other app installation and customization.

@miettal: the reason is simply, that we never manually create entries in Odoo, but instead automate this task via the Odoo API - and having the field read-only avoids nasty problems in production, with people randomly changing this critical value. Is there a way to have the field writable on creation and then read-only? 

We use a script like this to add devices:

```
#!/usr/bin/python3

# Requires
# apt install python3-odoorpc python3-dotenv

# Imports
import os
import random
import odoorpc
import sys
from dotenv import load_dotenv
from datetime import datetime

# Initialize config parameters
first_id = 1
last_id = 99999
producttype = ''

# For 'prod'
odoo_port = 8069
odoo_db = "prod"
odoo_host = "localhost"

# Load parameters from .env file
load_dotenv()
odoo_user = os.getenv("ODOO_USER")
odoo_password = os.getenv("ODOO_PASSWORD")

# Connect to Odoo
odoo = odoorpc.ODOO(odoo_host, port=odoo_port)
odoo.login(odoo_db, odoo_user, odoo_password)

Device = odoo.env['edge.device']

# Function to create passwords
def create_password(pw_length):
    alphabet = "abcdefghijklmnopqrstuvwxyz0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ!§$%&()"
    mypw = ""
    for i in range(pw_length):
        next_index = random.randrange(len(alphabet))
        mypw = mypw + alphabet[next_index]
    return mypw

# Loop over ids
for x in range(first_id, last_id + 1):
    name = "edge" + str(x)
    print("Create:", name)

    # Generate passwords
    apikey = create_password(160)

    # Create Odoo objects
    device = Device.create({
        'comment': name,
        'name': name,
        'apikey': apikey,
        'producttype': producttype
    })
```

> * The key for the configuration parameter "edge_monitoring_url" is wrong
> * the required name for device is readonly --> no creation is possible via UI
> * the computed field is not calculated correctly (TypeError: can only concatenate str (not "bool") to str)

@slash-ds: Thank you for the effort. I am surprised, that these changes are required, because I had tested this code successfully before committing it - and just now tested it again - on a fresh Gitpod environment -> https://gitpod.io/#https://github.com/OpenEMS/openems/tree/main

**Comment by miettal** [comment]
I see, I understood well. this is a guardrail for miss-operation. nice.

> Is there a way to have the field writable on creation and then read-only?

might be custom constraints is one solution. (I don't familiar with odoo system, I just googled it)

https://www.odoo.com/documentation/17.0/developer/tutorials/getting_started/11_constraints.html

**Comment by sfeilmeier** [comment]
Fixed in #7

**Comment by joern-schellhaas-itk** [comment]
I'm running odoo-openems@971ee74, and was not able to add a device name until I modified the `device.xml` as suggested here.

So, #7 does not fully fix the functionality for me.

**Pull request closed without merging**